### PR TITLE
Use `XDG Base Directory Specification` for configuration directory

### PR DESCRIPTION
**Pull request by rverst**
Implements #77

Uses the following strategy to get the configuration directory:

- look for `$XDG_CONFIG_HOME` and return `$XDG_CONFIG_HOME/bin` if set and exists
- look for `$HOME/.config` and return `$HOME/config/bin` if it exists
- as default return '$HOME/.bin'

### CODE REVIEW

**Comment by sirlatrom** [comment]
This will break existing users' configurations if they have the XDG variable and the existing config file

**Comment by rverst** [comment]
Good point, I've added a check for an existing configuration and return the default directory if a configuration exists.

File: pkg/config/config.go [hunk 1]
```diff
@@ -146,3 +146,36 @@ func GetOS() []string {
...
+	c := os.Getenv("XDG_CONFIG_HOME")
+	if _, err := os.Stat(c); !os.IsNotExist(err) {
+		return path.Join(c, "bin"), nil
+	}
+
+	if homeErr != nil {
+		return "", homeErr
+	}
+	c = path.Join(home, ".config")
+	if _, err := os.Stat(c); !os.IsNotExist(err) {
```
Comment (sirlatrom) [inline pkg/config/config.go:176]: What if `err != nil` otherwise?

File: pkg/config/config.go [hunk 2]
```diff
@@ -146,3 +146,36 @@ func GetOS() []string {
...
+func getConfigPath() (string, error) {
+	home, homeErr := os.UserHomeDir()
+	if homeErr == nil {
+		if _, err := os.Stat(path.Join(home, ".bin", "config.json")); !os.IsNotExist(err) {
+			return path.Join(path.Join(home, ".bin")), nil
+		}
+	}
+
+	c := os.Getenv("XDG_CONFIG_HOME")
+	if _, err := os.Stat(c); !os.IsNotExist(err) {
```
Comment (sirlatrom) [inline pkg/config/config.go:168]: What if `err != nil` otherwise?

File: pkg/config/config.go [hunk 3]
```diff
@@ -146,3 +146,36 @@ func GetOS() []string {
...
+//     exists and return "$HOME/.bin"
+//   - if "XDG_CONFIG_HOME" is set, return "$XDG_CONFIG_HOME/bin"
+//   - if "$HOME/.config" exists, return "$home/.config/bin"
+//   - default to "$HOME/.bin/"
+// ToDo: move the function to config_unix.go and add a similar function for windows,
+//       %APPDATA% might be the right place on windows
+func getConfigPath() (string, error) {
+	home, homeErr := os.UserHomeDir()
+	if homeErr == nil {
+		if _, err := os.Stat(path.Join(home, ".bin", "config.json")); !os.IsNotExist(err) {
```
Comment (sirlatrom) [inline pkg/config/config.go:162]: What if `err != nil` otherwise?

**Review by sirlatrom — CHANGES_REQUESTED** [review]
There may be other error reasons such as permissions errors.

File: pkg/config/config.go (hunk 3 above)
Comment (rverst) [reply pkg/config/config.go:162]: If `homeErr != nil` the check for `XDG_HOME_CONFIG` is done first, then the check for `homeErr` returns the error, if there is any.

Comment (rverst) [reply pkg/config/config.go:162]: Permissions errors will be handled as before in `CheckAndLoad()` or `write()`.

**Review by marcosnils — APPROVED** [review]
SGTM. after @sirlatrom approval we can merge.

File: pkg/config/config.go (hunk 3 above)
Comment (breml) [inline pkg/config/config.go:162]: Shouldn't this be `filepath.Join` as well (instead of `path.Join`)?
Also on the next line.

**Comment by breml** [comment]
Thanks for the PR. I like this, I look forward to have this merged.

**Comment by sirlatrom** [comment]
> SGTM. after @sirlatrom approval we can merge.

@marcosnils I'll give it a proper look tomorrow CET, want to make sure I understand the error handling. Or you can merge if you want, I can understand it later :smile:

**Comment by marcosnils** [comment]
> @marcosnils I'll give it a proper look tomorrow CET, want to make sure I understand the error handling. Or you can merge if you want, I can understand it later

No worries, we can wait. There's no need to rush it

**Pull request merged**